scene/gui/text_edit.cpp

```diff
@@ -1898,1662 +1898,1369 @@ int TextEdit::_calculate_spaces_till_next_right_indent(int column) {
+			if (k->is_action("ui_up")) {
+				if (completion_index > 0) {
+					completion_index--;
+				} else {
+					completion_index = completion_options.size() - 1;
 				}
-				if (wi == rows.size() - 1 || cursor.column == row_end_col) {
-					cursor_set_column(text[cursor.line].length());
+				completion_current = completion_options[completion_index];
+				update();
+				event_accepted = true;
+			}
+			if (k->is_action("ui_down")) {
+				if (completion_index < completion_options.size() - 1) {
+					completion_index++;
 				} else {
-					cursor_set_column(row_end_col);
+					completion_index = 0;
 				}
-
-				if (k->get_shift()) {
-					_post_shift_selection();
-				} else if (k->get_command() || k->get_control()) {
-					deselect();
+				completion_current = completion_options[completion_index];
+				update();
+				event_accepted = true;
+			}
+			if (k->is_action("ui_page_up")) {
+				completion_index -= get_theme_constant("completion_lines");
+				if (completion_index < 0) {
+					completion_index = 0;
 				}
-
-				_cancel_completion();
-				completion_hint = "";
-#endif
-			} break;
-			case KEY_KP_9: {
-				if (k->get_unicode() != 0) {
-					keycode_handled = false;
-					break;
+				completion_current = completion_options[completion_index];
+				update();
+				event_accepted = true;
+			}
+			if (k->is_action("ui_page_down")) {
+				completion_index += get_theme_constant("completion_lines");
+				if (completion_index >= completion_options.size()) {
+					completion_index = completion_options.size() - 1;
 				}
-				[[fallthrough]];
+				completion_current = completion_options[completion_index];
+				update();
+				event_accepted = true;
 			}
-			case KEY_PAGEUP: {
-				if (k->get_shift()) {
-					_pre_shift_selection();
+			if (k->is_action("ui_home")) {
+				if (completion_index > 0) {
+					completion_index = 0;
+					completion_current = completion_options[completion_index];
+					update();
 				}
-
-				int wi;
-				int n_line = cursor.line - num_lines_from_rows(cursor.line, get_cursor_wrap_index(), -get_visible_rows(), wi) + 1;
-				cursor_set_line(n_line, true, false, wi);
-
-				if (k->get_shift()) {
-					_post_shift_selection();
+				event_accepted = true;
+			}
+			if (k->is_action("ui_end")) {
+				if (completion_index < completion_options.size() - 1) {
+					completion_index = completion_options.size() - 1;
+					completion_current = completion_options[completion_index];
+					update();
 				}
-
+				event_accepted = true;
+			}
+			if (k->is_action("ui_accept")) {
+				_confirm_completion();
+				event_accepted = true;
+			}
+			if (k->is_action("ui_cancel")) {
 				_cancel_completion();
-				completion_hint = "";
+				event_accepted = true;
+			}
 
-			} break;
-			case KEY_KP_3: {
-				if (k->get_unicode() != 0) {
-					keycode_handled = false;
-					break;
+			// Handle Unicode here (if no modifiers active) and update autocomplete.
+			if (k->get_unicode() >= 32) {
+				if (!modifier_active && !readonly) {
```

Comment:
Needs to be `((!modifier_active || (!k->get_command() || k->get_command() && k->get_alt()) && !readonly)` for alt keys on windows and some keyboard layouts.

Reply:
Sorry, I don't understand the use case, could you clarify for me? So this allows unicode to be processed if Control+Alt is pressed. What usage does that have?

Reply:
`Control+Alt` == `AltGr` on windows, which allows writing special chars e.g. `>` on non-english keyboards, see #6851